ConstraintParser

s9latimm opened the pull request:
Adds a parser for extended constraint syntax.

resolves se-passau/VaRA#664

s9latimm committed:
Constraint lexer draft

A	include/vara/Feature/ConstraintParser.h
M	unittests/Feature/CMakeLists.txt
A	unittests/Feature/ConstraintParser.cpp

s9latimm committed:
Merge branch 'vara-dev' into f-ConstraintParser

M	include/vara/Feature/Constraint.h
M	include/vara/Feature/Feature.h
M	include/vara/Feature/Relationship.h
M	lib/Feature/FeatureModel.cpp
M	lib/Feature/FeatureModelWriter.cpp

s9latimm committed:
Fix test

M	unittests/Feature/ConstraintParser.cpp

vulder commented on include/vara/Feature/ConstraintParser.h:
```diff
@@ -0,0 +1,353 @@
+
+  [[nodiscard]] std::optional<std::string> getValue() { return Value; }
+
+  [[nodiscard]] Precedence calcPrecedence() const {
```
Precedence changes as discussed last Friday, switching default, what happens to brackets etc.

vulder commented on include/vara/Feature/ConstraintParser.h:
```diff
@@ -0,0 +1,353 @@
+  ConstraintToken(ConstraintTokenKind Kind) : Kind(Kind) {}
+  ConstraintToken(ConstraintTokenKind Kind, std::string Value)
+      : Kind(Kind), Value(Value) {}
+  ConstraintToken(const ConstraintToken &) = delete;
+  ConstraintToken &operator=(ConstraintToken &) = delete;
+  ConstraintToken(ConstraintToken &&) = delete;
+  ConstraintToken &operator=(ConstraintToken &&) = delete;
```
Why do we have to treat tokens as not copy/movable?

vulder commented on include/vara/Feature/ConstraintParser.h:
```diff
@@ -0,0 +1,353 @@
+
+class ConstraintLexer {
+public:
+  using TokenListTy = std::vector<std::unique_ptr<ConstraintToken>>;
```
```suggestion
  using TokenListTy = std::vector<ConstraintToken>;
```
Wo we really need a pointer indirection here? CTokens seem quite small, so a few copies/moves should not hurt.

vulder commented on include/vara/Feature/ConstraintParser.h:
```diff
@@ -0,0 +1,353 @@
+          ConstraintToken::ConstraintTokenKind::END_OF_FILE));
+    }
+
+    return std::move(TokenList);
```
```suggestion
    return TokenList;
```
If all return paths return the same local stack value we don't need to move it, RVO will take care of everything :smile: .

vulder commented on include/vara/Feature/ConstraintParser.h:
```diff
@@ -0,0 +1,353 @@
+private:
+  using ResultTy = std::pair<std::unique_ptr<ConstraintToken>, int>;
+
+  std::string Cnt;
```
I would move the member to the end of the class, makes them easier to find later on.

vulder commented on include/vara/Feature/ConstraintParser.h:
```diff
@@ -0,0 +1,353 @@
+  std::string Cnt;
+
+  ResultTy munch(const int Pos) {
+    // TODO(se-passau/VaRA#664)
```
What do you mean with this TODO?

vulder commented on include/vara/Feature/ConstraintParser.h:
```diff
@@ -0,0 +1,353 @@
+
+  ResultTy munch(const int Pos) {
+    // TODO(se-passau/VaRA#664)
+    switch (Cnt[Pos]) {
```
We should handle pattern groups where we can use ASCII offsets to determine the case beforehand.

vulder commented on include/vara/Feature/ConstraintParser.h:
```diff
@@ -0,0 +1,353 @@
+
+  std::string Cnt;
+
+  ResultTy munch(const int Pos) {
```
If I remember correctly, we wanted to call this `tokenize`?

vulder commented on include/vara/Feature/ConstraintParser.h:
```diff
@@ -0,0 +1,353 @@
+    case 'Z':
+      return munchIdentifier(Pos);
+    default:
+      return {std::move(std::make_unique<ConstraintToken>(
+                  ConstraintToken::ConstraintTokenKind::ERROR,
+                  std::string(1, Cnt[Pos]))),
+              1};
```
```suggestion
      return {std::make_unique<ConstraintToken>(
                  ConstraintToken::ConstraintTokenKind::ERROR,
                  std::string(1, Cnt[Pos])),
              1};
```
The created `unique_ptr` is already a rvalue so we don't need a move here.

vulder commented on include/vara/Feature/ConstraintParser.h:
```diff
@@ -0,0 +1,353 @@
+        break;
+      }
+    }
+    return {std::move(std::make_unique<ConstraintToken>(
+                ConstraintToken::ConstraintTokenKind::WHITESPACE)),
+            MunchLength};
```
```suggestion
    return {std::make_unique<ConstraintToken>(
                ConstraintToken::ConstraintTokenKind::WHITESPACE),
            MunchLength};
```

vulder commented on include/vara/Feature/ConstraintParser.h:
```diff
@@ -0,0 +1,353 @@
+    }
+  }
+
+  ResultTy munchWhitespace(const int Pos) {
```
You want to skip whitespace here or?

vulder commented on include/vara/Feature/ConstraintParser.h:
```diff
@@ -0,0 +1,353 @@
+
+  TokenListTy buildTokenList() {
+    TokenListTy TokenList;
+    for (int Pos = 0; Pos < Cnt.size();) {
```
Instead of manually tracking the Position in the text we could use a llvm::StringRef and update it when we need. Would allow us to use llvm parsing helpers, like skip whitespace or finding a specific next char....

vulder commented on include/vara/Feature/ConstraintParser.h:
```diff
@@ -0,0 +1,353 @@
+            MunchLength};
+  }
+
+  ResultTy munchOperator(const int Pos) {
```
munch munch :rofl: 
![munch](https://media.tenor.com/images/a2f5b1987aeec21bc546e85c7d4b3cef/tenor.gif)

vulder commented on include/vara/Feature/ConstraintParser.h:
```diff
@@ -0,0 +1,353 @@
+    }
+  }
+
+  ResultTy munchNumber(const int Pos) {
```
With StringRef this is a substring search + llvm::to_integer(), would make it a bit easier than handling everything our selfs.

vulder reviewed (changes requested):
Added a few suggestions

s9latimm committed:
Requested changes

M	include/vara/Feature/ConstraintParser.h
M	unittests/Feature/ConstraintParser.cpp

s9latimm committed:
Use StringRef

M	include/vara/Feature/ConstraintParser.h
